{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyNxxpYcX31cy3ol2cyXYgfZ",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/git4mamta/1_B-cell-Epitope-Prediction/blob/main/1_BIM2024_B_cell_Epitope_Prediction.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "y4mmJ1d79ls3",
        "outputId": "c0d65cd4-283d-41ea-84eb-99620f8eae66"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: biopython in /usr/local/lib/python3.12/dist-packages (1.85)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from biopython) (2.0.2)\n",
            "Requirement already satisfied: biopython in /usr/local/lib/python3.12/dist-packages (1.85)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from biopython) (2.0.2)\n",
            "Mounted at /content/drive\n"
          ]
        }
      ],
      "source": [
        "!pip install biopython\n",
        "!pip install biopython --upgrade\n",
        "\n",
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install biopython pandas\n",
        "!pip install biopython pandas --upgrade"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "AMaxlzMGK9j0",
        "outputId": "5734da93-eccc-4200-b210-de7def7b7947"
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: biopython in /usr/local/lib/python3.12/dist-packages (1.85)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.12/dist-packages (2.3.2)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from biopython) (2.0.2)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.12/dist-packages (from pandas) (2.9.0.post0)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.12/dist-packages (from pandas) (2025.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.12/dist-packages (from pandas) (2025.2)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.12/dist-packages (from python-dateutil>=2.8.2->pandas) (1.17.0)\n",
            "Requirement already satisfied: biopython in /usr/local/lib/python3.12/dist-packages (1.85)\n",
            "Requirement already satisfied: pandas in /usr/local/lib/python3.12/dist-packages (2.3.2)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from biopython) (2.0.2)\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.12/dist-packages (from pandas) (2.9.0.post0)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.12/dist-packages (from pandas) (2025.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.12/dist-packages (from pandas) (2025.2)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.12/dist-packages (from python-dateutil>=2.8.2->pandas) (1.17.0)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "amino_acid_properties = {\n",
        "    'A': {'hydropathy': 1.8, 'charge': 0, 'polarity': 0.16, 'flexibility': 0.357},\n",
        "    'R': {'hydropathy': -4.5, 'charge': 1, 'polarity': 0.16, 'flexibility': 0.529},\n",
        "    'N': {'hydropathy': -3.5, 'charge': 0, 'polarity': 0.16, 'flexibility': 0.463},\n",
        "    'D': {'hydropathy': -3.5, 'charge': -1, 'polarity': 0.16, 'flexibility': 0.511},\n",
        "    'C': {'hydropathy': 2.5, 'charge': 0, 'polarity': 0.16, 'flexibility': 0.346},\n",
        "    'Q': {'hydropathy': -3.5, 'charge': 0, 'polarity': 0.16, 'flexibility': 0.493},\n",
        "    'E': {'hydropathy': -3.5, 'charge': -1, 'polarity': 0.16, 'flexibility': 0.497},\n",
        "    'G': {'hydropathy': -0.4, 'charge': 0, 'polarity': 0.16, 'flexibility': 0.544},\n",
        "    'H': {'hydropathy': -3.2, 'charge': 0, 'polarity': 0.16, 'flexibility': 0.323},\n",
        "    'I': {'hydropathy': 4.5, 'charge': 0, 'polarity': 0.16, 'flexibility': 0.205},\n",
        "    'L': {'hydropathy': 3.8, 'charge': 0, 'polarity': 0.16, 'flexibility': 0.212},\n",
        "    'K': {'hydropathy': -3.9, 'charge': 1, 'polarity': 0.16, 'flexibility': 0.466},\n",
        "    'M': {'hydropathy': 1.9, 'charge': 0, 'polarity': 0.16, 'flexibility': 0.319},\n",
        "    'F': {'hydropathy': 2.8, 'charge': 0, 'polarity': 0.16, 'flexibility': 0.173},\n",
        "    'P': {'hydropathy': -1.6, 'charge': 0, 'polarity': 0.16, 'flexibility': 0.509},\n",
        "    'S': {'hydropathy': -0.8, 'charge': 0, 'polarity': 0.16, 'flexibility': 0.447},\n",
        "    'T': {'hydropathy': -0.7, 'charge': 0, 'polarity': 0.16, 'flexibility': 0.389},\n",
        "    'V': {'hydropathy': 4.2, 'charge': 0, 'polarity': 0.16, 'flexibility': 0.218},\n",
        "    'W': {'hydropathy': -0.9, 'charge': 0, 'polarity': 0.16, 'flexibility': 0.204},\n",
        "    'Y': {'hydropathy': -1.3, 'charge': 0, 'polarity': 0.16, 'flexibility': 0.298}\n",
        "}"
      ],
      "metadata": {
        "id": "-G4lr-TeLCTM"
      },
      "execution_count": 18,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio.SeqUtils import ProtParamData\n",
        "\n",
        "solv_accessibility = {\n",
        "    'A': 106.0,\n",
        "    'C': 135.0,\n",
        "    'D': 163.0,\n",
        "    'E': 194.0,\n",
        "    'F': 197.0,\n",
        "    'G': 84.0,\n",
        "    'H': 184.0,\n",
        "    'I': 169.0,\n",
        "    'K': 205.0,\n",
        "    'L': 164.0,\n",
        "    'M': 188.0,\n",
        "    'N': 157.0,\n",
        "    'P': 136.0,\n",
        "    'Q': 198.0,\n",
        "    'R': 248.0,\n",
        "    'S': 130.0,\n",
        "    'T': 142.0,\n",
        "    'V': 142.0,\n",
        "    'W': 227.0,\n",
        "    'Y': 222.0\n",
        "}\n",
        "\n",
        "ProtParamData.solv_accessibility = solv_accessibility"
      ],
      "metadata": {
        "id": "4YgZrd48LHo5"
      },
      "execution_count": 19,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "solv_accessibility"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Xejas545LOFr",
        "outputId": "87ab0894-f719-484b-b489-2179502d4597"
      },
      "execution_count": 20,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{'A': 106.0,\n",
              " 'C': 135.0,\n",
              " 'D': 163.0,\n",
              " 'E': 194.0,\n",
              " 'F': 197.0,\n",
              " 'G': 84.0,\n",
              " 'H': 184.0,\n",
              " 'I': 169.0,\n",
              " 'K': 205.0,\n",
              " 'L': 164.0,\n",
              " 'M': 188.0,\n",
              " 'N': 157.0,\n",
              " 'P': 136.0,\n",
              " 'Q': 198.0,\n",
              " 'R': 248.0,\n",
              " 'S': 130.0,\n",
              " 'T': 142.0,\n",
              " 'V': 142.0,\n",
              " 'W': 227.0,\n",
              " 'Y': 222.0}"
            ]
          },
          "metadata": {},
          "execution_count": 20
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "#from Bio.SeqUtils.PrimportotParam import ProteinAnalysis\n",
        "import numpy as np\n",
        "import io\n",
        "import math\n",
        "import joblib\n",
        "from joblib import load\n",
        "\n",
        "\n",
        "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
        "\n",
        "from Bio.SeqUtils import ProtParamData\n",
        "from Bio.SeqUtils import ProtParam\n",
        "from Bio.SeqUtils import MeltingTemp as mt\n",
        "from Bio.SeqUtils.ProtParamData import kd\n",
        "from Bio.SeqUtils.MeltingTemp import Tm_NN\n",
        "import Bio.SeqUtils.ProtParam as Bio\n",
        "from Bio.Seq import Seq\n",
        "import Bio.SeqUtils.ProtParamData as Bio\n",
        "from Bio.PDB.Polypeptide import is_aa\n",
        "from Bio.PDB import PDBParser, PPBuilder, Selection\n",
        "from Bio.PDB.vectors import calc_dihedral\n",
        "from Bio.PDB import PDBParser, PPBuilder, Polypeptide\n",
        "import warnings\n",
        "from Bio.PDB import*\n",
        "#from Bio.PDB.polypeptide import three_to_one\n",
        "from Bio.SeqUtils import seq1\n",
        "from Bio import SeqIO\n",
        "import Bio.PDB\n",
        "from Bio.PDB.Polypeptide import PPBuilder\n",
        "from Bio.Seq import Seq\n",
        "#from Bio.Alphabet import generic_protein\n",
        "#from Bio.SeqUtils.ProtParam import gravy\n",
        "\n",
        "\n",
        "df = pd.read_excel('/content/drive/My Drive/Colab Notebooks/Linear_B-cell.xlsx')\n",
        "absorbance = 1.5\n",
        "path_length = 1\n",
        "concentration = 1e-6\n",
        "\n",
        "valid_aas = set(\"ACDEFGHIKLMNPQRSTVWY\")  # standard amino acids\n",
        "df['Sequence'] = df['Sequence'].apply(lambda s: ''.join([aa for aa in s if aa in valid_aas]))\n",
        "\n",
        "def calculate_hydrophilicity(sequence):\n",
        "    protein = ProteinAnalysis(sequence)\n",
        "    gravy = protein.gravy()\n",
        "    return abs(gravy)\n",
        "\n",
        "def calculate_surface_accessibility(sequence):\n",
        "    protein = ProteinAnalysis(sequence)\n",
        "    accessibility = protein.flexibility()\n",
        "    surface_accessibility = sum([accessibility[i] for i, aa in enumerate(sequence) if aa in ProtParamData.solv_accessibility and ProtParamData.solv_accessibility[aa] > 0.05 and i < len(accessibility)]) / len(sequence)\n",
        "    return surface_accessibility\n",
        "\n",
        "def calculate_beta_turn(sequence):\n",
        "    protein = ProteinAnalysis(sequence)\n",
        "    beta_turn = protein.secondary_structure_fraction()[2]\n",
        "    return beta_turn\n",
        "\n",
        "def calculate_exposed_surface(sequence):\n",
        "    protein = ProteinAnalysis(sequence)\n",
        "    exposed_surface = sum([1 for i, aa in enumerate(sequence) if ProtParamData.solv_accessibility[aa] > 0]) / len(sequence)\n",
        "    return exposed_surface\n",
        "\n",
        "def calculate_polarity(sequence):\n",
        "    protein = ProtParam.ProteinAnalysis(sequence)\n",
        "    amino_acids = protein.count_amino_acids()\n",
        "    polar_aa = ['R', 'N', 'D', 'Q', 'E', 'H', 'K', 'S', 'T', 'Y']\n",
        "    polar_count = sum(amino_acids[aa] for aa in polar_aa)\n",
        "    average_polarity = polar_count / len(sequence)\n",
        "    return average_polarity\n",
        "\n",
        "def calculate_antigenic(sequence):\n",
        "    protein = ProteinAnalysis(sequence)\n",
        "    antigenic = [amino_acid_properties[aa]['flexibility'] for aa in sequence]\n",
        "    average_antigenic = sum(antigenic) / len(sequence)\n",
        "    return average_antigenic\n",
        "\n",
        "def calculate_molecular_weight(protein_sequence):\n",
        "    protein = ProteinAnalysis(protein_sequence)\n",
        "    mw = protein.molecular_weight()\n",
        "    return mw\n",
        "\n",
        "def calculate_hydrophobicity(sequence):\n",
        "    protein = ProtParam.ProteinAnalysis(sequence)\n",
        "    hydrophobicity = protein.protein_scale(window=9, param_dict=kd, edge=1)\n",
        "    return hydrophobicity\n",
        "\n",
        "def calculate_immunogenicity(sequence):\n",
        "    antigenicity_scale = {\n",
        "        'A': 0.8, 'R': 1.0, 'N': 0.8, 'D': 0.9, 'C': 0.8, 'Q': 0.9,\n",
        "        'E': 1.0, 'G': 0.8, 'H': 0.9, 'I': 0.8, 'L': 0.8, 'K': 1.0,\n",
        "        'M': 0.9, 'F': 0.8, 'P': 0.8, 'S': 0.8, 'T': 0.8, 'W': 0.8,\n",
        "        'Y': 0.8, 'V': 0.8\n",
        "    }\n",
        "    immunogenicity = sum(antigenicity_scale.get(aa, 0) for aa in sequence) / len(sequence)\n",
        "    return immunogenicity\n",
        "\n",
        "def calculate_sequence_length(sequence):\n",
        "    return len(sequence)\n",
        "\n",
        "def calculate_electrostatic_interaction(sequence):\n",
        "    # Calculate net charge of the protein sequence using ProteinAnalysis class\n",
        "    protein = ProteinAnalysis(sequence)\n",
        "    net_charge = protein.charge_at_pH(7.4)\n",
        "    return net_charge\n",
        "\n",
        "def calculate_flexibility(sequence):\n",
        "    protein = ProteinAnalysis(sequence)\n",
        "    flexibility = protein.flexibility()\n",
        "    return sum(flexibility) / len(sequence)\n",
        "\n",
        "def calculate_pI(sequence):\n",
        "     protein_analysis = ProteinAnalysis(sequence)\n",
        "\n",
        "     pKa_values = protein_analysis.isoelectric_point()\n",
        "\n",
        "     if isinstance(pKa_values, float):\n",
        "        n = 1\n",
        "        pI = pKa_values\n",
        "     else:\n",
        "        n = len(pKa_values)\n",
        "        pI = sum(pKa_values) / n\n",
        "\n",
        "     return pI\n",
        "\n",
        "def calculate_half_life(sequence):\n",
        "    F = 0.7\n",
        "    t = 10\n",
        "\n",
        "    num_cleavable_bonds = sequence.count('D') + sequence.count('E') + sequence.count('R') + sequence.count('K')\n",
        "    length = len(sequence)\n",
        "    num_bonds_per_unit_length = num_cleavable_bonds / length\n",
        "\n",
        "    if num_bonds_per_unit_length == 0:\n",
        "        return None\n",
        "\n",
        "    kdeg = (1 - F) / (t * F * num_bonds_per_unit_length)\n",
        "\n",
        "    ln2 = math.log(2)\n",
        "    half_life = 1 / (kdeg * ln2)\n",
        "\n",
        "    return half_life\n",
        "\n",
        "\n",
        "def calculate_aliphatic_index(sequence):\n",
        "\n",
        "    rel_vol = {'A': 1.00, 'V': 2.90, 'I': 3.10, 'L': 4.90}\n",
        "\n",
        "\n",
        "    num_a = sequence.count('A')\n",
        "    num_v = sequence.count('V')\n",
        "    num_i = sequence.count('I')\n",
        "    num_l = sequence.count('L')\n",
        "\n",
        "\n",
        "    sum_rel_vol = num_a * rel_vol['A'] + num_v * rel_vol['V'] + num_i * rel_vol['I'] + num_l * rel_vol['L']\n",
        "\n",
        "\n",
        "    n = len(sequence)\n",
        "    aliphatic_index = sum_rel_vol / n\n",
        "\n",
        "    return aliphatic_index\n",
        "\n",
        "def calculate_molecular_extinction_coefficient(sequence, absorbance, path_length, concentration):\n",
        "\n",
        "\n",
        "    n = len(sequence)\n",
        "    molecular_extinction_coefficient = absorbance / (path_length * concentration * n)\n",
        "\n",
        "    return molecular_extinction_coefficient\n",
        "\n",
        "\n",
        "def calculate_aromaticity(sequence):\n",
        "\n",
        "    if len(sequence) % 2 != 0:\n",
        "        return False\n",
        "\n",
        "    num_pi_electrons = 0\n",
        "    for atom in sequence:\n",
        "        if atom in [\"C\", \"N\"]:\n",
        "            num_pi_electrons += 1\n",
        "\n",
        "    # the Hückel rule\n",
        "    if num_pi_electrons % 4 == 2:\n",
        "        return True\n",
        "    else:\n",
        "        return False\n",
        "\n",
        "def calculate_virulence(sequence):\n",
        "    virulent_amino_acids = ['G', 'K', 'R', 'F', 'Y']\n",
        "\n",
        "    virulent_count = 0\n",
        "    for aa in sequence:\n",
        "        if aa in virulent_amino_acids:\n",
        "            virulent_count += 1\n",
        "\n",
        "    virulence_score = virulent_count / len(sequence)\n",
        "\n",
        "    return virulence_score\n",
        "\n",
        "def calculate_instability_index(sequence):\n",
        "    unstable_dipeptides = ['GA', 'GS', 'GT', 'AV', 'VL', 'VA', 'IA', 'PA', 'PS', 'NP']\n",
        "\n",
        "    unstable_count = 0\n",
        "    for i in range(len(sequence) - 1):\n",
        "        dipeptide = sequence[i:i+2]\n",
        "        if dipeptide in unstable_dipeptides:\n",
        "            unstable_count += 1\n",
        "\n",
        "    instability_index = 10 * unstable_count / len(sequence)\n",
        "\n",
        "    return instability_index\n",
        "\n",
        "def calculate_adhesin_probability(sequence):\n",
        "    adhesin_motifs = ['RGD', 'RGDS', 'KGD', 'NPLY', 'FNxGNxD', 'GGxxG', 'GxxG', 'DxxD']\n",
        "\n",
        "    adhesin_count = 0\n",
        "    for motif in adhesin_motifs:\n",
        "        if motif in sequence:\n",
        "            adhesin_count += 1\n",
        "\n",
        "    adhesin_probability = adhesin_count / len(adhesin_motifs)\n",
        "\n",
        "    return adhesin_probability\n",
        "\n",
        "def calculate_df(sequence):\n",
        "\n",
        "    def_matrix = {\n",
        "        'A': 0.326,\n",
        "        'C': 0.346,\n",
        "        'D': 0.572,\n",
        "        'E': 0.511,\n",
        "        'F': 0.314,\n",
        "        'G': 0.544,\n",
        "        'H': 0.323,\n",
        "        'I': 0.308,\n",
        "        'K': 0.525,\n",
        "        'L': 0.310,\n",
        "        'M': 0.324,\n",
        "        'N': 0.454,\n",
        "        'P': 0.511,\n",
        "        'Q': 0.497,\n",
        "        'R': 0.548,\n",
        "        'S': 0.509,\n",
        "        'T': 0.444,\n",
        "        'V': 0.308,\n",
        "        'W': 0.305,\n",
        "        'Y': 0.420\n",
        "    }\n",
        "\n",
        "    def_energy = [def_matrix[aa] for aa in sequence]\n",
        "\n",
        "    df = sum(def_energy) / len(sequence)\n",
        "\n",
        "    return df\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "'''def predict_allergenicity(sequence):\n",
        "    model = joblib.load('allergenicity_model.joblib')\n",
        "\n",
        "    # Convert the protein sequence to a feature vector\n",
        "    # (e.g., using a k-mer or amino acid composition encoding)\n",
        "    feature_vector = convert_sequence_to_features(sequence)\n",
        "\n",
        "    # Make a prediction using the machine learning model\n",
        "    prediction = model.predict(np.array([feature_vector]))\n",
        "\n",
        "    # Convert the prediction to a boolean value\n",
        "    if prediction == 1:\n",
        "        return True\n",
        "    else:\n",
        "        return False\n",
        "        '''\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "'''def calculate_dihedral_angles(fasta_seq):\n",
        "    # Load sequence into BioPython structure\n",
        "    parser = Bio.PDB.PDBParser()\n",
        "    seq_id = \"SEQ\"\n",
        "    sequence = Bio.PDB.Polypeptide.Polypeptide(fasta_seq, id=seq_id)\n",
        "\n",
        "    # Calculate dihedral angles\n",
        "    ppb = PPBuilder()\n",
        "    structure = parser.get_structure(seq_id, sequence)\n",
        "    model = structure[0]\n",
        "    pp = ppb.build_peptides(model)\n",
        "    angles = []\n",
        "    for chain in pp:\n",
        "        dihedral_list = Bio.PDB.Polypeptide.calc_dihedrals(chain)\n",
        "        angles.append(dihedral_list)\n",
        "\n",
        "    return angles'''\n",
        "\n",
        "\n",
        "'''def calculate_stability(sequence):\n",
        "    \"\"\"\n",
        "    Calculates the stability of a protein sequence based on its melting temperature\n",
        "    and grand average of hydropathy (GRAVY) score.\n",
        "    \"\"\"\n",
        "    # Calculate the melting temperature using the nearest-neighbor method\n",
        "    # with the same parameters as before\n",
        "    seq = Seq(sequence)\n",
        "    try:\n",
        "        tm = Tm_NN(seq, Na=50, Mg=1.5, dNTPs=0.8, saltcorr=7)\n",
        "    except ZeroDivisionError:\n",
        "        return np.nan\n",
        "\n",
        "    # Calculate the grand average of hydropathy (GRAVY) score\n",
        "    gravy = Bio.gravy(sequence)\n",
        "\n",
        "    # Calculate the stability score as the product of the melting temperature\n",
        "    # and the absolute value of the GRAVY score\n",
        "    stability = abs(tm * gravy)\n",
        "\n",
        "    return stability'''\n",
        "\n",
        "\n",
        "\n",
        "df['Hydrophilicity'] = df['Sequence'].apply(calculate_hydrophilicity)\n",
        "df['Surface Accessibility'] = df['Sequence'].apply(calculate_surface_accessibility)\n",
        "df['Beta-turn'] = df['Sequence'].apply(calculate_beta_turn)\n",
        "df['Exposed Surface'] = df['Sequence'].apply(calculate_exposed_surface)\n",
        "df['Polarity'] =df['Sequence'].apply(calculate_polarity)\n",
        "df['Antigenic'] =df['Sequence'].apply(calculate_antigenic)\n",
        "df['Molecular Weight'] = df['Sequence'].apply(calculate_molecular_weight)\n",
        "df['Hydrophobicity'] = df['Sequence'].apply(calculate_hydrophobicity)\n",
        "df['Immunogenicity'] = df['Sequence'].apply(calculate_immunogenicity)\n",
        "df['Sequence Length'] = df['Sequence'].apply(calculate_sequence_length)\n",
        "df['Electrostatic Interaction'] = df['Sequence'].apply(calculate_electrostatic_interaction)\n",
        "df['Instability index'] = df['Sequence'].apply(calculate_instability_index)\n",
        "#df['Stability'] = df['Sequence'].apply(calculate_stability)\n",
        "df['Flexibility'] = df['Sequence'].apply(calculate_flexibility)\n",
        "df['pI'] = df['Sequence'].apply(calculate_pI)\n",
        "df['Estimated Half Life'] = df['Sequence'].apply(calculate_half_life)\n",
        "df['Aliphatic Index'] = df['Sequence'].apply(calculate_aliphatic_index)\n",
        "df['Aromacity'] = df['Sequence'].apply(calculate_aromaticity)\n",
        "df['Virulance'] = df['Sequence'].apply(calculate_virulence)\n",
        "df['Deformability'] = df['Sequence'].apply(calculate_df)\n",
        "df['Adhesin probability'] = df['Sequence'].apply(calculate_adhesin_probability)\n",
        "#df['Allergenicity'] = df['Sequence'].apply(predict_allergenicity)\n",
        "df['Molecular Extinction Coefficient'] = df['Sequence'].apply(calculate_molecular_extinction_coefficient, args=(absorbance, path_length, concentration))\n",
        "'''for index, row in df.iterrows():\n",
        "    # Get the protein sequence\n",
        "    sequence = row['Sequence']\n",
        "\n",
        "    # Create a FASTA-formatted string\n",
        "    fasta_str = '>{}\\n{}'.format(index, sequence)\n",
        "\n",
        "    # Pass the FASTA-formatted sequence to a function\n",
        "    #your_function(fasta_str)\n",
        "    angles = calculate_dihedral_angles(fasta_str)\n",
        "    print(angles)'''\n",
        "\n",
        "df.to_excel('b_cell_epitopes_with_features.xlsx', index=False)"
      ],
      "metadata": {
        "id": "cAzD-z_rLdrZ"
      },
      "execution_count": 16,
      "outputs": []
    }
  ]
}